em/emResidualFunctionals.cc

```diff
@@ -0,0 +1,932 @@
+    PCU_ALWAYS_ASSERT(up.n == 1);
+    apf::MeshEntity* firstTet = up.e[0];
+    tets.push_back(firstTet);
+
+    apf::MeshEntity* nextFace = getTetOppFaceSharingEdge(mesh, firstTet,
+        firstFace, edge);
+    apf::MeshEntity* nextTet = firstTet;
+    mesh->getUp(nextFace, up);
+    while( up.n == 2) {
+      faces.push_back(nextFace);
+      if (nextTet != up.e[0])
+        nextTet = up.e[0];
+      else
+        nextTet = up.e[1];
+      tets.push_back(nextTet);
+
+      nextFace = getTetOppFaceSharingEdge(mesh, nextTet, nextFace, edge);
+      mesh->getUp(nextFace, up);
+    }
+    faces.push_back(nextFace);
+  }
+}
+
+void getClockwiseTetsandFaces(EdgePatch* p)
+{
+  if (p->isOnBdry) {
+    if (p->tets.size() > 1) {
+      apf::MeshEntity* secondFace = p->faces[1];
+      apf::MeshEntity* firstTet = p->tets[0];
+      apf::MeshEntity* secondTet = p->tets[1];
+
+      apf::Downward firstTetFaces, secondTetFaces;
+      int n_firstTetFaces = p->mesh->getDownward(firstTet, 2, firstTetFaces);
+      int n_secondTetFaces = p->mesh->getDownward(secondTet, 2, secondTetFaces);
+      int fi1 = apf::findIn(firstTetFaces, n_firstTetFaces, secondFace);
+      int fi2 = apf::findIn(secondTetFaces, n_secondTetFaces, secondFace);
+      PCU_ALWAYS_ASSERT(fi1 != -1 && fi2 != -1);
+
+      // first tet opp vertex crd
+      apf::MeshEntity* firstTetOppVert = getTetOppVert(
+        p->mesh, firstTet, secondFace);
+      apf::Vector3 firstTetOppVertCrd;
+      p->mesh->getPoint(firstTetOppVert, 0, firstTetOppVertCrd);
+
+      // second tet opp vertex crd
+      apf::MeshEntity* secondTetOppVert = getTetOppVert(
+        p->mesh, secondTet, secondFace);
+      apf::Vector3 secondTetOppVertCrd;
+      p->mesh->getPoint(secondTetOppVert, 0, secondTetOppVertCrd);
+
+      // normal to the face
+      apf::Downward edge_vertices;
+      p->mesh->getDownward(p->entity, 0, edge_vertices);
+      apf::Vector3 p0, p1, p2;
+      p->mesh->getPoint(edge_vertices[0], 0, p0);
+      p->mesh->getPoint(edge_vertices[1], 0, p1);
+      apf::MeshEntity* faceOppVert = getFaceOppVert(
+        p->mesh, secondFace, p->entity);
+      p->mesh->getPoint(faceOppVert, 0, p2);
+
+      apf::Vector3 normal = apf::cross(p1-p0, p2-p0);
+
+      // direction vectors from p0 to opp tet verts
+      apf::Vector3 vFirst = firstTetOppVertCrd - p0;
+      apf::Vector3 vLast = secondTetOppVertCrd - p0;
+
+      if ((vFirst * normal > 0)  && (vLast * normal < 0)) {
+        // reverse list of tets and faces
+        std::reverse(p->tets.begin(), p->tets.end());
+        std::reverse(p->faces.begin(), p->faces.begin());
+      }
+      else if ((vFirst * normal < 0)  && (vLast * normal > 0)) {
+        cout << "already clockwise" << endl;
+      }
+      else if ((vFirst * normal < 0)  && (vLast * normal < 0)) {
+        cout << "failed clockwise. ABORT" << endl;
+      }
+      else if ((vFirst * normal > 0)  && (vLast * normal > 0)) {
+        cout << "failed clockwise. ABORT" << endl;
```

Comment:
Use `PCU_ALWAYS_ASSERT_VERBOSE` instead.

Reply:
These were just used for debugging purposes, so i have removed them.